### PR TITLE
feat: migration for FIP-0034

### PR DESCRIPTION
**Pull request by arajasek**
The specification for the change can be found [here](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0034.md), and the Rust-based implementation can be found [here](https://github.com/filecoin-project/builtin-actors/pull/484).

### CODE REVIEW

**Review by anorth — COMMENTED** [review]
Thanks for this!

**Comment by jennijuju** [comment]
https://github.com/filecoin-project/go-commp-utils/pull/9

File: builtin/v9/migration/top.go [hunk 1]
```diff
@@ -140,8 +150,35 @@ func MigrateStateTree(ctx context.Context, store cbor.IpldStore, newManifestCID
...
+	if !ok {
+		return cid.Undef, xerrors.New("didn't find market actor")
+	}
+
+	var marketState market8.State
+	if err := store.Get(ctx, marketActor.Head, &marketState); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get system actor state: %w", err)
+	}
+
+	proposals, err := market8.AsDealProposalArray(adtStore, marketState.Proposals)
```
Comment (arajasek) [inline builtin/v9/migration/top.go:175]: I need someone to talk me through thread-safety here. I'm pretty sure this _isn't_ safe. 

If so, do we need to load the proposals AMT each time we migrate a miner [with deals in their pending precommits]? That wouldn't be too bad (ultimately not _that_ many such loads), but I'm wondering if people have better ideas.

File: builtin/v9/miner/miner_types.go [hunk 2]
```diff
@@ -53,6 +53,21 @@ type ProveReplicaUpdatesParams struct {
...
+	Partition            uint64
+	NewSealedSectorCID   cid.Cid `checked:"true"`
+	NewUnsealedSectorCID cid.Cid `checked:"true"`
+	Deals                []abi.DealID
+	UpdateProofType      abi.RegisteredUpdateProof
+	ReplicaProof         []byte
+}
+
+type ProveReplicaUpdatesParams2 struct {
+	Updates []ReplicaUpdate
```
[the remark this answers is not in the document]
Comment (arajasek) [reply builtin/v9/miner/miner_types.go]: Fixed, thanks for the catch

File: builtin/v9/migration/miner.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	newHead, err := store.Put(ctx, &outState)
+	return &actorMigrationResult{
+		newCodeCID: m.migratedCodeCID(),
+		newHead:    newHead,
+	}, err
+}
+
+func generateUnsealedCID(proofType abi.RegisteredSealProof, pieces []abi.PieceInfo) (cid.Cid, error) {
```
Comment (arajasek) [inline builtin/v9/migration/miner.go:133]: This method became a lot smaller after I took out the padding logic (which the commp util already does). I'll flatten it into the caller.

File: builtin/v9/migration/miner.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+		err = newPrecommitOnChainInfos.Put(miner9.SectorKey(info.Info.SectorNumber), &miner9.SectorPreCommitOnChainInfo{
+			Info: miner9.SectorPreCommitInfo{
+				SealProof:     info.Info.SealProof,
+				SectorNumber:  info.Info.SectorNumber,
+				SealedCID:     info.Info.SealedCID,
+				SealRandEpoch: info.Info.SealRandEpoch,
+				DealIDs:       info.Info.DealIDs,
+				Expiration:    info.Info.Expiration,
+				UnsealedCid:   unsealedCid,
```
Comment (ribasushi) [inline builtin/v9/migration/miner.go:86]: All looks good. Idly wondering why is this is a `*cid.Cid` as opposed to `cid.Cid`. Is it to allow for the optional/assumed CC CommD?

File: builtin/v9/migration/miner.go (hunk 3 above)
Comment (ribasushi) [reply builtin/v9/migration/miner.go:133]: They are like 1:1 identical signatures now no?

**Review by geoff-vball — APPROVED** [review]

File: builtin/v9/migration/miner.go (hunk 3 above)
Comment (arajasek) [reply builtin/v9/migration/miner.go:133]: Yes they are.

**Pull request merged**